src/tree/updater_quantile_hist.cc

```diff
@@ -777,45 +775,44 @@ void QuantileHistMaker::Builder::EvaluateSplits(const std::vector<ExpandEntry>&
 // on comparison of indexes values (idx_span) and split point (split_cond)
 // Handle dense columns
 // Analog of std::stable_partition, but in no-inplace manner
-template <bool default_left>
+template <bool default_left, typename BinIdxType>
 inline std::pair<size_t, size_t> PartitionDenseKernel(
-      common::Span<const size_t> rid_span, common::Span<const uint32_t> idx_span,
-      const int32_t split_cond, const uint32_t offset,
-      common::Span<size_t> left_part, common::Span<size_t> right_part) {
-  const uint32_t* idx = idx_span.data();
+      common::Span<const size_t> rid_span, common::Span<const BinIdxType> idx_span,
+      const int32_t split_cond, const int32_t offset,
+      common::Span<size_t> left_part, common::Span<size_t> right_part,
+      const std::vector<bool>* missing_val_flag, const size_t disp) {
+  const BinIdxType* idx = idx_span.data();
   size_t* p_left_part = left_part.data();
   size_t* p_right_part = right_part.data();
   size_t nleft_elems = 0;
   size_t nright_elems = 0;
 
-  const uint32_t missing_val = std::numeric_limits<uint32_t>::max();
-
   for (auto rid : rid_span) {
-    if (idx[rid] == missing_val) {
+    if ((*missing_val_flag)[disp + rid]) {
```

Comment:
Can we the the column in this kernel, and add a function to `Column` named `IsMissing(size_t rid)`?  I don't feel comfortable passing around pointers.

Reply:
Done :)